### PR TITLE
Fix Machine ID selection of Kubernetes Access address/port

### LINKED ISSUE #19811: Machine ID Kubernetes Access incorrectly uses web address/port when multiplexing disabled
### Expected behavior:

machine id should match the tsh behavior when generating the host:port for use in a kubeconfig when multiplexing is disabled.

### Current behavior:

Given a cluster with multiplexing disabled and no explicit `kube_public_addr` set, tsh kube login and machine id with a kubeconfig destination will not use the same host:port for the same cluster.

The `tsh kube login` version will generate a `cluster:` section with `server: https://cluster.example.com:3026`, which is correct. The kubernetes service is not enabled on the web port in this cluster configuration.

If a kubeconfig is generated with a machine id bot, even when using the same exact roles and kube cluster, the resulting kubeconfig ends up with `server: https://cluster.example.com:3080`. Since there is no kube service running on port 3080, kubectl gives a certificate error. The cert error is incidental-- the kubeconfig is expecting a cert signed by the Teleport CA, but the web proxy service does not do that. Even if one did adjust the kubeconfig accordingly, the web proxy port is still not running the kube service when multiplexing is disabled.

```
$ curl https://cluster.example.com:3080/webapi/ping
{"auth":{"type":"local","second_factor":"off","local":{"name":""},"private_key_policy":"none","has_motd":false},"proxy":{"kube":{"enabled":true,"listen_addr":"0.0.0.0:3026"},"ssh":{"listen_addr":"[::]:3023","tunnel_listen_addr":"0.0.0.0:3024","web_listen_addr":"0.0.0.0:3080","public_addr":"cluster.example.com:3080","ssh_public_addr":"cluster.example.com:3023","ssh_tunnel_public_addr":"cluster.example.com:3024"},"db":{},"tls_routing_enabled":false},"server_version":"11.1.4","min_client_version":"10.0.0","cluster_name":"cluster"}
$ export KUBECONFIG=~/.kube/config
$ tsh login --user testuser --proxy cluster.example.com:3080
$ tsh kube login default
$ grep server: $KUBECONFIG
    server: https://cluster.example.com:3026
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.0", GitCommit:"a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2", GitTreeState:"clean", BuildDate:"2022-08-23T17:44:59Z", GoVersion:"go1.19", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.10", GitCommit:"eae22ba6238096f5dec1ceb62766e97783f0ba2f", GitTreeState:"clean", BuildDate:"2022-05-24T12:50:52Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
$ tbot start -c /opt/bot/bot.cfg &
$ export KUBECONFIG=/opt/bot/dest/kubeconfig.yaml
$ grep server: $KUBECONFIG
    server: https://cluster.example.com:3080
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.0", GitCommit:"a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2", GitTreeState:"clean", BuildDate:"2022-08-23T17:44:59Z", GoVersion:"go1.19", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v4.5.7
Unable to connect to the server: x509: certificate signed by unknown authority
```

### Workarounds:

The machine id user would need to set up some type of process to adjust the kubeconfig.yaml output on the fly.

_or_

The Teleport admin could set `kube_public_addr` explicitly to match the expected hostname/port rather than relying on the behavior of the tsh/tctl client binaries to correctly use the correct value.

### Bug details:
- Teleport version - reported against 9.3.24, also present in `Teleport v11.1.4 git:v11.1.4-0-ge4ac5f6717 go1.19.4`
- Recreation steps - set up a cluster with multiplexing disabled, kube access enabled, at least one kube cluster. Omit the `kube_public_addr` in the teleport.yaml for the proxy. compare the generated kubeconfig files for both `tsh kube login` and `tbot start` and observe that machine id is using the web port incorrectly.
- Debug logs - n/a

Here's an example teleport.yaml that was used with 11.1.4 to recreate the issue locally:

```yaml
version: v3
teleport:
  nodename: recreate
  data_dir: /opt/recreate/teleport
  log:
    output: stderr
    severity: INFO
    format:
      output: text
  ca_pin: ""
  diag_addr: ""
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  cluster_name: "cluster"
  license_file: /opt/recreate/teleport/license.pem
  proxy_listener_mode: separate
  authentication:
    type: local
    second_factor: off
ssh_service:
  enabled: "no"
proxy_service:
  enabled: "yes"
  web_listen_addr: 0.0.0.0:3080
  public_addr: cluster.example.com:5443
  listen_addr: "[::]:3023"
  tunnel_listen_addr: 0.0.0.0:3024
  ssh_public_addr: cluster.example.com:3023
  tunnel_public_addr: cluster.example.com:3024
  kube_listen_addr: 0.0.0.0:3026
  #kube_public_addr: cluster.example.com:3026
  https_keypairs: []
  acme: {}

kubernetes_service:
  enabled: "yes"
  kubeconfig_file: /opt/recreate/kind_kubeconfig
  listen_addr: 0.0.0.0:3027
```

### LINKED ISSUE #34178: Unable to use dedicated kube port when tls routing is enabled
Expected behavior:
1. tbot should prefer to use the multiplex tls port (443) when tls routing is enabled even when a dedicated kube port is available.
2. The teleport proxy should be able to handle kubernetes client connections to the dedicated kube port regardless of whether tls routing is enabled.

Current behavior:
1. tbot uses the dedicated kube port when tls routing is enabled.
2. Teleport proxy does not properly handle connections on the dedicated kube port when tls routing is enabled.

Bug details:
- Teleport version

Observed on Teleport Cloud v14.1.1. Works properly on Teleport Cloud v13.4.5.
tbot version 14.1.1

- Recreation steps
1. Configure proxy_listener_mode: multiplex on the auth service
```yaml
auth_service:
...
proxy_listener_mode: multiplex
...
```

3. Configure a dedicated kube_public_addr/kube_listener_addr in the proxy service
```yaml
proxy_service:
...
kube_listen_addr: 0.0.0.0:3026
kube_public_addr: example.teleport.sh:11057
...
```

4. Configure tbot to generate a kubernetes config
```yaml
version: v2

debug: true
auth_server: "example.teleport.sh:443"
certificate_ttl: "1h"

renewal_interval: "1m"

oneshot: false
onboarding:
  token: "xxxxxx"
  join_method: "token"
storage:
  type: directory
  path: ./tbot/data

outputs:
  - type: kubernetes
    kubernetes_cluster: cloud-docker-minikube
    destination:
      type: directory
      path: ./tbot/kube
```
5. Observe ./tbot/kube/kubeconfig.yaml is configured to point at the dedicated port.

```yaml
server: https://example.teleport.sh:11387
tls-server-name: kube-teleport-proxy-alpn.example.teleport.sh
```

6. Observe running a kubectl command using the kubeconfig from tbot fails.
`kubectl version --kubeconfig kubeconfig.yaml`

- Debug logs
```bash
kubectl version --kubeconfig kubeconfig.yaml
Unable to connect to the server: EOF
```

### PR DESCRIPTION
**Pull request by strideynet**
Closes https://github.com/gravitational/teleport/issues/19811 /  Semi-closes https://github.com/gravitational/teleport/issues/34178

Addresses a few long standing issues w/ the selection of what address to use when tbot generates the `kubeconfig` file. This also feels a bit neater/more explicit than the previous code and I've also included more tests to cover this process since there's a few edgecases.

changelog: Fixes issue where `tbot` would select the wrong address for Kubernetes Access when in ports separate mode.

### CODE REVIEW

**Comment by strideynet** [comment]
Backport will exclude the docs changes ([b23d53a](https://github.com/gravitational/teleport/pull/34202/commits/b23d53a9c7a6b401f01ce6e7a3ccc9b828078d49)) and a separate PR will be raised when this change is released to include a note specifying the version this is fixed within.

File: lib/tbot/config/template_kubernetes.go [hunk 1]
```diff
@@ -226,3 +183,51 @@ func (t *templateKubernetes) render(
...
+		addr := proxyPong.Proxy.SSH.PublicAddr
+		host, _, err := net.SplitHostPort(proxyPong.Proxy.SSH.PublicAddr)
+		if err != nil {
+			return "", "", trace.Wrap(err, "parsing proxy public_addr")
+		}
+
+		sni = fmt.Sprintf("%s%s", constants.KubeTeleportProxyALPNPrefix, host)
+		hostIsIP := net.ParseIP(host) != nil
+		if host == "" || hostIsIP {
+			sni = fmt.Sprintf("%s%s", constants.KubeTeleportProxyALPNPrefix, constants.APIDomain)
```
Comment (zmb3) [inline lib/tbot/config/template_kubernetes.go:205]:
```suggestion
			sni = constants.KubeTeleportProxyALPNPrefix+ constants.APIDomain
```

Using `Sprintf` to concatenate 2 strings is kind of overkill, no?

**Review by zmb3 — APPROVED** [review]

**Review by ptgott — APPROVED** [review]

File: lib/tbot/config/template_kubernetes.go (hunk 1 above)
Comment (tigrato) [reply lib/tbot/config/template_kubernetes.go:205]: Just use `client.GetKubeTLSServerName`

Comment (strideynet) [reply lib/tbot/config/template_kubernetes.go]: Good shout @tigrato - 8bbff66eb2

**Review by tigrato — APPROVED** [review]

**Pull request merged**